core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1751,4 +1752,18 @@ public static RelDataType extractLastNFields(RelDataTypeFactory typeFactory,
     return typeFactory.createStructType(
         type.getFieldList().subList(fieldsCnt - numToKeep, fieldsCnt));
   }
+
+  public static boolean isValidDecimalValue(BigDecimal value, RelDataType toType) {
+    if (value == null) {
+      return true;
+    }
+    switch (toType.getSqlTypeName()) {
+    case DECIMAL:
+      final int intDigits = value.precision() - value.scale();
+      final int maxIntDigits = toType.getPrecision() - toType.getScale();
+      return intDigits <= maxIntDigits;
```

Comment:
Seems the `toType` should always be a `Decimal` type, we should add a check here. Also, can we add some doc to this method ?

Reply:
> Seems the toType should always be a Decimal type, we should add a check here

`toType` can be any type. I used `case when`. So I think it is fine.

> Also, can we add some doc to this method ?

Makes sense.

Reply:
If `toType` is not decimal, how can you ensure that it has a valid precision and scale, such as `-1`.

Reply:
Do you want to throw an exception if toType is not decimal by saying "add a check here"?

Reply:
I just want to make sure the logic is robust, either a check or assert is okey from  my side.